Exiled.CustomRoles/API/Features/CustomRole.cs

```diff
@@ -0,0 +1,445 @@
+
+            customRoles = tempList.AsReadOnly();
+            ListPool<CustomRole>.Shared.Return(tempList);
+
+            return customRoles?.Count > 0;
+        }
+
+        /// <summary>
+        /// Checks if the given player has this role.
+        /// </summary>
+        /// <param name="player">The <see cref="Player"/> to check.</param>
+        /// <returns>True if the player has this role.</returns>
+        public virtual bool Check(Player player) => TrackedPlayers.Contains(player);
+
+        /// <summary>
+        /// Tries to register this role.
+        /// </summary>
+        /// <returns>True if the role registered properly.</returns>
+        public bool TryRegister()
+        {
+            if (!Registered.Contains(this))
+            {
+                if (Registered.Any(r => r.Id == Id))
+                {
+                    Log.Warn($"{Name} has tried to register with the same Role ID as another role: {Id}. It will not be registered!");
+
+                    return false;
+                }
+
+                Registered.Add(this);
+                Init();
+
+                Log.Debug($"{Name} ({Id}) has been successfully registered.", CustomRoles.Instance.Config.Debug);
+
+                return true;
+            }
+
+            Log.Warn($"Couldn't register {Name} ({Id}) [{Role}] as it already exists.");
+
+            return false;
+        }
+
+        /// <summary>
+        /// Tries to unregister this role.
+        /// </summary>
+        /// <returns>True if the role is unregistered properly.</returns>
+        public bool TryUnregister()
+        {
+            Destroy();
+
+            if (!Registered.Remove(this))
+            {
+                Log.Warn($"Cannot unregister {Name} ({Id}) [{Role}], it hasn't been registered yet.");
+
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Initializes this role manager.
+        /// </summary>
+        public virtual void Init() => SubscribeEvents();
+
+        /// <summary>
+        /// Destroys this role manager.
+        /// </summary>
+        public virtual void Destroy() => UnSubscribeEvents();
+
+        /// <summary>
+        /// Handles setup of the role, including spawn location, inventory and registering event handlers.
+        /// </summary>
+        /// <param name="player">The <see cref="Player"/> to add the role to.</param>
+        public virtual void AddRole(Player player)
+        {
+            Log.Debug($"{Name}: Adding role to {player.Nickname}.", CustomRoles.Instance.Config.Debug);
+            if (Role != RoleType.None)
+            {
+                player.SetRole(Role, SpawnReason.ForceClass, true);
+            }
+
+            Timing.CallDelayed(1.5f, () =>
+            {
+                Vector3 pos = GetSpawnPosition();
+                if (pos != Vector3.zero)
+                    player.Position = pos + (Vector3.up * 1.5f);
```

Comment:
Please explain.

Reply:
Making sure they don't spawn inside the floor.